apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+      return Promise.reject(
+        new Error(`Couldn't find view ${viewName}`)
+      );
+    }
+
+    if (this.isCurrentPanel(viewName, args)) {
+      debug('View %s is the current view.', viewName);
+      return Promise.resolve();
     }
 
-    var index;
-    if ((index = hash.indexOf('&')) !== -1) {
-      hash = hash.slice(0, index);
+    var hash = '#';
+    if (view.partOf) {
+      hash += '/' + viewName;
     }
+    hash = Utils.url(hash, args);
+
+    return startNavigation(
+      { newPanel: view, args, partial: true }
+    ).then(
+      () => executeNavigationStep('beforeLeave')
+    ).then(() => {
+      var nextLocation;
+      if (navigationTransition.leaveDocument) {
+        var mainView = findMainView(view);
+        nextLocation = '/' + mainView.url + hash;
+      } else {
+        nextLocation = hash;
+      }
 
-    return hash;
+      setLocation(nextLocation);
+    }).then(
+      () => navigationTransition.defer.promise,
+      onNavigationError
+    );
+    // TODO do something on pageshow?
+    //
   },
 
-  slide: function n_slide(position) {
-    var wrapper = this.mainWrapper;
+  /**
+   * This is called when a view is started.
+   *
+   * It tries to find the right view from the location, and then initiates a
+   * navigation from this.
+   */
+  init() {
+    debug('init()');
+
+    attachHistoryListener();
 
-    if (wrapper.dataset.position === position) {
+    var newPanel = findViewFromLocation(window.location);
+    if (!newPanel) {
+      console.error('Couldn\'t find a view from location', window.location);
       return Promise.resolve();
     }
 
-    wrapper.dataset.position = position;
+    var args = Utils.params(decodeURIComponent(window.location.hash));
+
+    debug('init: found panel', newPanel.name, 'with args', args);
+
+    return startNavigation({newPanel, args}).then(
+      prepareSwitchPanel
+    ).then(
+      // right away as we don't execute anything on the previous panel, and we
+      // need a state at startup.
+      setFutureState
+    ).then(
+      () => executeNavigationStep('beforeEnter').catch(catchStepError)
+    ).then(
+      switchPanel
+    ).then(
+      attachAfterEnterHandler
+    ).then(
+      endNavigation,
+      onNavigationError
+    );
+  },
 
-    return new Promise(function(resolve, reject) {
-      // We have 2 panels, so we get 2 transitionend for each step
-      var trEndCount = 0;
+  /* will be used by tests */
+  cleanup() {
+    detachHistoryListener();
+  },
 
-      wrapper.addEventListener('transitionend', function trWait(e) {
-        trEndCount++;
+  /**
+   * Checks if current location.hash corresponds to default panel. This is
+   * sometimes called before we resolve the state, in that case resolve
+   * directly using findViewFromHash. Otherwise we use the resolved state.
+   *
+   * @returns {boolean} True if current location.hash corresponds to default
+   * panel.
+   */
+  isDefaultPanel() {
+    var state = currentState;
```

Comment:
I think I did it at one moment when the `? :` expression could fit in one line. But now it can't anyway.